docs/proposals/20231122-automate-aks-features.md

```diff
@@ -0,0 +1,423 @@
+new AKS features. There would continue to be some amount of ongoing cost though to identify and handle nuances
+for each feature though, which may include issues outside of CAPZ's control like AKS API quirks.
+
+The main drawback of this approach from a developer perspective is the up-front effort to implement and
+ongoing cost to maintain the code generation itself. The existing AKS feature set exposed by CAPZ provides a
+decent foundation that can help identify regressions by testing the pipeline against features already
+implemented and tested. ASO also already implements a full code generation pipeline to transform Azure API
+specs into Kubernetes resource definitions, so some of that could possibly be reused by CAPZ.
+
+#### Option 7: CAPZ resource defines patches to ASO resource
+
+This method describes adding a new `spec.asoManagedClusterPatches` field to the existing
+AzureManagedControlPlane:
+
+```go
+type AzureManagedControlPlaneClassSpec struct {
+	...
+
+	// ASOManagedClusterPatches defines patches to be applied to the generated ASO ManagedCluster resource.
+	ASOManagedClusterPatches []string `json:"asoManagedClusterPatches,omitempty"`
+}
+```
+
+After CAPZ calculates the ASO ManagedCluster for a given AzureManagedControlPlane during a reconciliation, it
+will apply these patches in order to the ManagedCluster before ultimately submitting that to ASO. This allows
+users to specify any AKS feature declaratively within the existing CAPZ spec. The exact format of the patches
+is TBD, but likely one or all of the formats described here:
+https://kubernetes.io/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch/
+
+The main drawback of this approach is the fragility of the patches themselves, which can break underneath
+users with a new version of ASO or CAPZ. There is also increased risk that a user could modify the ASO
+resource in a way that breaks CAPZ's ability to reconcile it, though CAPZ could perhaps add some extra
+validation that particularly sensitive fields do not get modified by a set of patches. Compared to other
+options, the syntax for specifying a patch is also more cumbersome than if its equivalent had its own CAPZ API
+field. Given the intent of setting patches is to enable niche use-cases for advanced users, these drawbacks
+may be acceptable.
+
+#### Option 8: Users bring-their-own ASO ManagedCluster resource
+
+CAPZ can already incorporate existing ASO resources into Cluster API Cluster configurations. BYO ASO resources
+are never modified or deleted by CAPZ directly but are still read and play in to the status of CAPZ resources.
+
+The flow for this method is roughly:
+1. User creates an ASO ManagedCluster before or at the same time as the other Cluster API and CAPZ resources.
+1. Any updates to the AKS cluster are done by the user through the ASO resource.
+1. The user may control whether or not the ASO ManagedCluster gets deleted with the rest of the Cluster API
```

Comment:
So basically. As a user, I submit:

```yaml
...
---
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: AzureManagedControlPlane
...
---
apiVersion: containerservice.azure.com/v1api20230201storage
kind: ManagedCluster
...
```

And that indicates "let CAPZ create the resource group, and delete the ASO resource later when the parent Cluster is deleted".

But if I submit this:

```yaml
...
---
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: AzureManagedControlPlane
...
---
apiVersion: containerservice.azure.com/v1api20230201storage
kind: ManagedCluster
...
---
apiVersion: containerservice.azure.com/v1api20230201storage
kind: ResourceGroup
...
```

That would indicate that the user wants to self-manage the resource group, as well as any non-CAPZ-specific resources inside, after the CAPI Cluster resource is deleted?

Reply:
Correct on both. To clarify, CAPZ would still create and manage other resources it needs to inside a user-provided resource group (e.g. vnet) if those are also not BYO.